### PR TITLE
Gh10001 Compute taxes based on C_Invoice.M_Warehouse_ID

### PR DESCRIPTION
**Pull request by metas-rc**
https://github.com/metasfresh/metasfresh/issues/10001

### CODE REVIEW

File: backend/de.metas.adempiere.adempiere/base/src/main/java-gen/org/compiere/model/I_C_Invoice.java [hunk 1]
```diff
@@ -1971,4 +1971,33 @@
...
     /** Column name UserFlag */
     public static final String COLUMNNAME_UserFlag = "UserFlag";
+    
+    
+    /**
+	 * Set Lager.
+	 * Storage Warehouse and Service Point
+	 *
+	 * <br>Type: TableDir
+	 * <br>Mandatory: true
```
Comment (teosarca) [inline backend/de.metas.adempiere.adempiere/base/src/main/java-gen/org/compiere/model/I_C_Invoice.java:1981]: why is the Warehouse mandatory?

File: backend/de.metas.business/src/main/java/de/metas/invoice/service/IInvoiceBL.java [hunk 2]
```diff
@@ -423,4 +424,6 @@ void registerLineCopyHandler(
 	 * - set Price_UOM_ID to C_InvoiceLine.C_UOM_ID
 	 */
 	void ensureUOMsAreNotNull(@NonNull InvoiceId invoiceId);
+
+	void setInvoiceLineTaxes(Properties ctx, de.metas.adempiere.model.@NonNull I_C_Invoice invoice);
```
Comment (teosarca) [inline backend/de.metas.business/src/main/java/de/metas/invoice/service/IInvoiceBL.java:428]: why `ctx` is needed. For legacy APIs u can extract it from `invoice`.

File: backend/de.metas.business/src/main/java/de/metas/invoice/service/IInvoiceLineBL.java [hunk 3]
```diff
@@ -98,4 +104,6 @@
 	void updateLineNetAmt(I_C_InvoiceLine line, BigDecimal qtyEntered);
 
 	void updatePrices(I_C_InvoiceLine invoiceLine);
+
+	boolean setTaxForInvoiceLine(Properties ctx, org.compiere.model.I_C_InvoiceLine il, OrgId orgId, Timestamp taxDate, CountryId countryFromId, BPartnerLocationId partnerLocationId, boolean isSOTrx);
```
Comment (teosarca) [inline backend/de.metas.business/src/main/java/de/metas/invoice/service/IInvoiceLineBL.java:108]: why `ctx` is needed. for legacy APIs u can extract it from `il`.

File: backend/de.metas.business/src/main/java/de/metas/invoice/service/impl/AbstractInvoiceBL.java [hunk 4]
```diff
@@ -1371,6 +1383,66 @@ public final boolean isTaxIncluded(@NonNull final org.compiere.model.I_C_Invoice
 		return invoice.isTaxIncluded(); // 08486: use the invoice's flag, not whatever the PL sais right now
 	}
 
+	@Override
+	public void setInvoiceLineTaxes(final Properties ctx, @NonNull final I_C_Invoice invoice)
```
Comment (teosarca) [inline backend/de.metas.business/src/main/java/de/metas/invoice/service/impl/AbstractInvoiceBL.java:1387]: the `fromCountryId` in this method is computed in a different way compared with MInvoiceLine.setTax() method.
Pls make sure we are always computing things in one way.

More, I would advise to introduce some getFromCountryId(I_C_Invoice invoice, I_C_InvoiceLine invoiceLine) method in IInvoiceBL.
You could call that method from both places.

File: backend/de.metas.business/src/main/java/de/metas/invoice/service/impl/InvoiceLineBL.java [hunk 5]
```diff
@@ -108,6 +114,68 @@ public void setTaxAmtInfo(final Properties ctx, final I_C_InvoiceLine il, final
...
+		final Timestamp shipDate = io.getMovementDate();
+
+		final BPartnerLocationId partnerLocationId = BPartnerLocationId.ofRepoId(io.getC_BPartner_ID(), io.getC_BPartner_Location_ID());
+
+		final WarehouseId warehouseId = WarehouseId.ofRepoId(io.getM_Warehouse_ID());
+
+		final boolean isSOTrx = io.isSOTrx();
+
+		final CountryId countryFromId;
+		final WarehouseId invoiceWarehouseId = WarehouseId.ofRepoIdOrNull(invoice.getM_Warehouse_ID());
```
Comment (teosarca) [inline backend/de.metas.business/src/main/java/de/metas/invoice/service/impl/InvoiceLineBL.java:146]: avoid duplicating the get `countryFromId` logic.
see above.

File: backend/de.metas.business/src/main/java/de/metas/invoice/service/impl/InvoiceLineBL.java [hunk 6]
```diff
@@ -108,6 +114,68 @@ public void setTaxAmtInfo(final Properties ctx, final I_C_InvoiceLine il, final
...
+		else
+		{
+			countryFromId = warehouseBL.getCountryId(warehouseId);
+		}
+
+		return setTaxForInvoiceLine(ctx, il, inOutOrgId, shipDate, countryFromId, partnerLocationId, isSOTrx);
+	}
+
+	@Override
+	public boolean setTaxForInvoiceLine(final Properties ctx,
```
Comment (teosarca) [inline backend/de.metas.business/src/main/java/de/metas/invoice/service/impl/InvoiceLineBL.java:161]: minor: avoid `ctx` param

File: backend/de.metas.business/src/main/java/de/metas/invoice/service/impl/InvoiceLineBL.java [hunk 7]
```diff
@@ -108,6 +114,68 @@ public void setTaxAmtInfo(final Properties ctx, final I_C_InvoiceLine il, final
...
+			final org.compiere.model.I_C_InvoiceLine il,
+			final OrgId orgId,
+			final Timestamp taxDate,
+			final CountryId countryFromId,
+			final BPartnerLocationId partnerLocationId,
```
Comment (teosarca) [inline backend/de.metas.business/src/main/java/de/metas/invoice/service/impl/InvoiceLineBL.java:166]: rename `partnerLocationId` parameter to `billLocationId` or similar (to underline that we are talking about BILL BPartner and location)

Comment (teosarca) [reply backend/de.metas.business/src/main/java/de/metas/invoice/service/impl/InvoiceLineBL.java:166]: ... or is this the shipToLocation? can u pls clarify?

**Review by teosarca — CHANGES_REQUESTED** [review]
check my comments

File: backend/de.metas.adempiere.adempiere/base/src/main/java-gen/org/compiere/model/I_C_Invoice.java (hunk 1 above)
Comment (metas-rc) [reply backend/de.metas.adempiere.adempiere/base/src/main/java-gen/org/compiere/model/I_C_Invoice.java]: Copy - paste problem.
Regenerated the classes and hey are now ok.

File: backend/de.metas.business/src/main/java/de/metas/invoice/service/IInvoiceBL.java (hunk 2 above)
Comment (metas-rc) [reply backend/de.metas.business/src/main/java/de/metas/invoice/service/IInvoiceBL.java:428]: ✔️

File: backend/de.metas.business/src/main/java/de/metas/invoice/service/impl/InvoiceLineBL.java (hunk 7 above)
Comment (metas-rc) [reply backend/de.metas.business/src/main/java/de/metas/invoice/service/impl/InvoiceLineBL.java:154]: Hi, @teosarca 
* I checked and saw that before, we took the shipment bp and location whensetting the tax from ```de.metas.invoice.service.impl.InvoiceLineBL.setTax(I_C_InvoiceLine, String)``` and the invoice bp and location when setting it from ```org.compiere.model.MInvoiceLine.setTax()```
* Note that in the latter one the inout is not even searched for.
* Which partner and location should be considered when computing the tax? How is this legally done in the real world? 
* Similar: Which date should be used on tax computation? The shipment date or the invoice date?

Comment (metas-rc) [reply backend/de.metas.business/src/main/java/de/metas/invoice/service/impl/InvoiceLineBL.java:154]: I renamed the method to make it obvious the data is based on the inoutLine linked to the invoice. If there is no inoutline, the tax is not set.

**Pull request merged**